challenge(formatter): (also parser) Respect, parse, and format BOM characters from source files

faultyserver opened the pull request:
<!--
	Thanks for submitting a Pull Request! We appreciate you spending the time to work on these changes.
	Please provide enough information so that others can review your PR.
	Once created, your PR will be automatically labeled according to changed files.
	Learn more about contributing: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md
-->

See the end of this PR description for a note on reviewing this PR. The GitHub file viewer will not like it if you try to open it as a whole...

## Summary

About a dozen of the tests in the prettier suite use files that include a Unicode BOM character. Prettier's support for BOM characters is simply reading a file, stripping the BOM if it sees one, then doing all the work, and finally re-adding the BOM at the start of the result if it found one earlier. That works, but isn't easily adaptable into Biome as a more comprehensive tool.

BOM characters are a way of indicating which encoding a file is using, but it's also not a perfect mechanism. So much so that Unicode has generally stopped recommending the use of BOM characters [unless really necessary, which is almost never](https://www.unicode.org/L2/L2021/21038-bom-guidance.pdf). In UTF-8 files, the BOM generally has no meaning anyway, since byte ordering does not matter. Windows (and specifically Windows text editors like Visual Studio) however, for a variety of reasons, will often encode UTF-8 with a BOM, because [historically, Windows used UTF-16LE](https://superuser.com/a/294221), so it tries to disambiguate.

The presence of a BOM character generally has no meaning and no effect on tools, other than occasionally breaking them when the tool doesn't know that a BOM character might show up at the start of a file. But for exactly that reason (to avoid breaking on accident), it is worth respecting the presence of a BOM character, both when parsing and when formatting.

When parsing, the BOM can be read and stored alongside the parse tree. Then, when formatting, whatever BOM is included with the tree gets formatted back out at the start of the result. This ensures that BOMs are preserved when given, and not added or removed unintentionally.

A fun tangent in my explorations of this revealed that [JSON explicitly forbids BOM characters](https://datatracker.ietf.org/doc/html/rfc8259#section-8.1.) (when transmitting JSON over the network). This would imply that, as a technicality, allowing a BOM character while parsing JSON would be invalid. However, again for the sake of consistency with other tools and to avoid breaking unintentionally (if, say, a Windows users creates a JSON file that inadvertently gets a BOM character when saved to a file), I think it makes sense to respect the BOM and just let it pass by like the other languages.

## Approach

It took me a while to figure out how it would even be feasible to handle BOMs across all languages and ways that text could be sent to Biome to be parsed, linted, and formatted. Originally, I wanted to do it as abstractly as possible; in one place, high up when first reading text from a file. But, this quickly seemed like a rough idea, especially since the formatting methods for creating format results don't actually operate on files, but purely on a parse result. I thought about adding it as state in the Lexer as well, just seeing if the stream starts with a BOM, and then other components can use that state, but that also didn't seem feasible since the Lexer is really an internal detail of parsing, and formatters don't directly know about the Lexer right now.

So the BOM data needs to be passed along with the tree no matter what, which led to adding it as a token in the root nodes of each language tree. This has the unfortunate side effect that _every_ parse will have a slot for the BOM token, which will _almost always_ be empty, but that's also true with the interpreter token right now, so I felt it was an acceptable means to the end here.

If there is an easier, less intrusive way of going about this, I'm all for it lol.

One other benefit of parsing a BOM token and including it in the AST is that Biome can also use that in lint rules to warn against the usage of BOMs. For example, it could raise an error in JSON files that have a BOM token, since that is almost always going to be invalid for various tools. Also, if any BOM other than plain UTF-8 is encountered, it could warn that the encoding doesn't match what Biome expects.


<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve?-->

<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->

## Test Plan

To review this PR, I've split it up into a few specific commits. The first commit should contain all of the logic changes. Everything with adding the `unicode_bom` crate to each parser, implementing the `UNICODE_BOM` token kind, lexing it, storing it in the parse nodes, and writing them out in the formatter from the root nodes of each language. This is really the only interesting thing in this PR, and it's much smaller.

The second and third commits are just updating the `.rast` test snapshots for all of the spec tests to include the bom_token slot on each node.

The fourth commit is the removal of the prettier test snapshots for the tests that are now passing completely. See the last commit for all of the removed tests. They are:
- [typescript/compiler/castParentheses.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptcompilercastparenthesests)
- [typescript/compiler/commentInNamespaceDeclarationWithIdentifierPathName.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptcompilercommentinnamespacedeclarationwithidentifierpathnamets)
- [typescript/conformance/es6/templates/templateStringWithEmbeddedTypeAssertionOnAdditionES6.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancees6templatestemplatestringwithembeddedtypeassertiononadditiones6ts)
- [typescript/conformance/types/tuple/contextualTypeWithTuple.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypestuplecontextualtypewithtuplets)
- [typescript/conformance/types/tuple/indexerWithTuple.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypestupleindexerwithtuplets)
- [typescript/conformance/types/tuple/typeInferenceWithTupleType.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypestupletypeinferencewithtupletypets)
- [typescript/conformance/types/union/unionTypeCallSignatures.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypesunionuniontypecallsignaturests)
- [typescript/conformance/types/union/unionTypeCallSignatures3.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypesunionuniontypecallsignaturests3)
- [typescript/conformance/types/union/unionTypeCallSignatures4.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypesunionuniontypecallsignaturests4)
- [typescript/conformance/types/union/unionTypeConstructSignatures.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypesunionuniontypeconstructsignaturests)
- [typescript/conformance/types/union/unionTypeEquivalence.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypesunionuniontypeequivalencets)
- [typescript/conformance/types/union/unionTypeFromArrayLiteral.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypesunionuniontypefromarrayliteralts)
- [typescript/conformance/types/union/unionTypeIndexSignature.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptconformancetypesunionuniontypeindexsignaturets)

That these tests now pass (and all of the others as well), should indicate that BOMs are being respected in files where they are present and not affecting parsing or formatting unintentionally.

ematipico commented on crates/biome_css_parser/Cargo.toml:
```diff
@@ -19,6 +19,7 @@ biome_js_unicode_table = { workspace = true }
 biome_parser           = { workspace = true }
 biome_rowan            = { workspace = true }
 tracing                = { workspace = true }
+unicode-bom = "2.0.3"
```
Since we are about to use this crate in multiple crates, we usually add a dependency at the root `Cargo.toml`, and then here you'll have

```toml 
unicode-bom = { workspace = true }
```

faultyserver replied:
i didn't realize this was a feature! that's pretty neat. i'll update it.

ematipico commented on crates/biome_css_parser/src/lexer/mod.rs:
```diff
@@ -259,6 +267,27 @@ impl<'src> CssLexer<'src> {
+    /// Must be called at a valid UT8 char boundary (and realistically only at
+    /// the start position of the source).
+    fn consume_potential_bom(&mut self) -> Option<CssSyntaxKind> {
+        if let Some(first) = self.source().get(0..3) {
```
Can we add a comment that tells us why we require inspecting a slice from `0` to `3`? I'm sure there's a reason!

ematipico reviewed (approved):
The solution looks good to me! Thank you very much for the thorough explanation! I am sure people learned a lot from this PR (included myself)

Conaclos reviewed (approved):
This is a valuable contribution!

faultyserver commented:
Sorry for the many pushes and build notifications, but I think this has addressed the feedback, fixed the tests, and cleaned up some unused changes i'd made originally before going in the final direction here. I think it's set to go now.
